File: scripts/sync_registry.js
import fs from 'node:fs';
import path from 'node:path';
import { execSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT_DIR = path.resolve(__dirname, '..');
const DATA_DIR = path.join(ROOT_DIR, 'server', 'data');
const REGISTRY_DIR = path.join(DATA_DIR, 'registry');
const ZIP_PATH = path.join(DATA_DIR, 'registry_master.zip');
const EXTRACT_TMP = path.join(DATA_DIR, 'registry_tmp_extract');

export async function downloadAndSyncOfficialRegistry() {
  if (!fs.existsSync(DATA_DIR)) {
    fs.mkdirSync(DATA_DIR, { recursive: true });
  }

  const archiveUrl = 'https://git.lantian.pub/backup/dn42-registry/archive/master.zip';
  console.log(`🌐 [DN42 Registry] Downloading full registry archive from ${archiveUrl} ...`);

  const res = await fetch(archiveUrl, { headers: { 'User-Agent': 'AkiLab-DN42-Portal/2.0' } });
  if (!res.ok) {
    throw new Error(`Download failed with status ${res.status}`);
  }

  const buffer = Buffer.from(await res.arrayBuffer());
  fs.writeFileSync(ZIP_PATH, buffer);
  console.log(`📦 [DN42 Registry] Downloaded ${(buffer.length / 1024 / 1024).toFixed(2)} MB archive.`);

  if (fs.existsSync(EXTRACT_TMP)) {
    fs.rmSync(EXTRACT_TMP, { recursive: true, force: true });
  }
  fs.mkdirSync(EXTRACT_TMP, { recursive: true });

  console.log('📂 [DN42 Registry] Extracting objects via PowerShell Expand-Archive...');
  // Note: Windows NTFS does not support colons in filenames (inet6num/route6).
  // Expand-Archive might warn or skip colons, but aut-num, mntner, person, and role contain NO colons and extract 100% cleanly!
  try {
    execSync(`powershell -NoProfile -Command "Expand-Archive -LiteralPath '${ZIP_PATH}' -DestinationPath '${EXTRACT_TMP}' -Force"`, { stdio: 'ignore' });
  } catch {
    // Ignore NTFS colon filename errors on inet6num
    console.log('ℹ️ Extracted core objects (skipped NTFS invalid colon filenames).');
  }

  const rootItems = fs.readdirSync(EXTRACT_TMP);
  const innerFolder = rootItems.length > 0 ? path.join(EXTRACT_TMP, rootItems[0]) : EXTRACT_TMP;
  const sourceDataDir = path.join(innerFolder, 'data');

  if (fs.existsSync(REGISTRY_DIR)) {
    fs.rmSync(REGISTRY_DIR, { recursive: true, force: true });
  }
  fs.mkdirSync(REGISTRY_DIR, { recursive: true });

  const destDataDir = path.join(REGISTRY_DIR, 'data');
  fs.mkdirSync(destDataDir, { recursive: true });

  // Copy aut-num, mntner, person, role directories
  const targetDirs = ['aut-num', 'mntner', 'person', 'role'];
  for (const dirName of targetDirs) {
    const src = path.join(sourceDataDir, dirName);
    const dest = path.join(destDataDir, dirName);
    if (fs.existsSync(src)) {
      fs.cpSync(src, dest, { recursive: true });
      console.log(`  ✓ Copied ${dirName} (${fs.readdirSync(dest).length} objects)`);
    }
  }

  console.log(`✅ [DN42 Registry] Successfully installed official registry objects to ${destDataDir}`);

  // Cleanup temp files
  try { fs.unlinkSync(ZIP_PATH); } catch {}
  try { fs.rmSync(EXTRACT_TMP, { recursive: true, force: true }); } catch {}

  // Verify AS4242421337
  const testAsPath = path.join(destDataDir, 'aut-num', 'AS4242421337');
  if (fs.existsSync(testAsPath)) {
    console.log('\n🔍 [Verification] Verified AS4242421337 object from official registry:');
    console.log(fs.readFileSync(testAsPath, 'utf-8'));
  }
}

downloadAndSyncOfficialRegistry().catch((err) => {
  console.error('❌ Registry sync error:', err.message);
});

File: scripts/sync_registry.py
import urllib.request
import zipfile
import io
import os
import shutil
import sys

# Force UTF-8 output on Windows terminal
if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8")
    sys.stderr.reconfigure(encoding="utf-8")

ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DEST_REGISTRY_DIR = os.path.join(ROOT_DIR, "server", "data", "registry", "data")

TARGET_PREFIXES = ("data/aut-num/", "data/mntner/", "data/person/", "data/role/")
ARCHIVE_URL = "https://git.lantian.pub/backup/dn42-registry/archive/master.zip"

def sync():
    print(f"[DN42 Registry] Downloading official registry zip from {ARCHIVE_URL}...")
    req = urllib.request.Request(ARCHIVE_URL, headers={"User-Agent": "AkiLab-DN42-Portal/2.0"})
    with urllib.request.urlopen(req) as resp:
        data = resp.read()
    print(f"[DN42 Registry] Downloaded {len(data) / (1024*1024):.2f} MB in memory.")

    if os.path.exists(DEST_REGISTRY_DIR):
        shutil.rmtree(DEST_REGISTRY_DIR)
    os.makedirs(DEST_REGISTRY_DIR, exist_ok=True)

    extracted_count = 0
    with zipfile.ZipFile(io.BytesIO(data)) as z:
        for member in z.infolist():
            parts = member.filename.split("/", 1)
            rel_path = parts[1] if len(parts) > 1 else member.filename

            if any(rel_path.startswith(prefix) for prefix in TARGET_PREFIXES) and not member.is_dir():
                dest_file_path = os.path.join(os.path.dirname(DEST_REGISTRY_DIR), rel_path.replace("/", os.sep))
                os.makedirs(os.path.dirname(dest_file_path), exist_ok=True)
                with z.open(member) as src, open(dest_file_path, "wb") as dst:
                    dst.write(src.read())
                extracted_count += 1

    print(f"[DN42 Registry] Successfully extracted {extracted_count} registry objects to {DEST_REGISTRY_DIR}!")

    # Verify AS4242423143
    test_as_file = os.path.join(DEST_REGISTRY_DIR, "aut-num", "AS4242423143")
    if os.path.exists(test_as_file):
        print("\n[Verification] Verified AS4242423143:")
        with open(test_as_file, "r", encoding="utf-8") as f:
            print(f.read())

    test_mnt_file = os.path.join(DEST_REGISTRY_DIR, "mntner", "AKIRA-MNT")
    if os.path.exists(test_mnt_file):
        print("\n[Verification] Verified AKIRA-MNT:")
        with open(test_mnt_file, "r", encoding="utf-8") as f:
            print(f.read())

if __name__ == "__main__":
    sync()
